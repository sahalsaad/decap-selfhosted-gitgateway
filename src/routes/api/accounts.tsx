import { faker } from '@faker-js/faker'
import { zValidator } from '@hono/zod-validator'
import { hashPassword } from '@services/encryption-service'
import { UserService } from '@services/user-service'
import { Hono } from 'hono'
import { jwtAdminMiddleware, jwtMiddleware } from '@/src/middlewares/jwt'
import { resetPasswordRequestSchema, setPasswordRequestSchema } from '@/types/password'
import type { Variables } from '@/types/variables'

const accountsRoute = new Hono<{
  Bindings: CloudflareBindings
  Variables: Variables
}>()

accountsRoute.post(
  '/reset-password',
  jwtMiddleware,
  jwtAdminMiddleware,
  zValidator('json', resetPasswordRequestSchema),
  async (ctx) => {
    const jsonBody = ctx.req.valid('json')
    const userService = UserService(ctx.env.DB, ctx.env.AUTH_SECRET_KEY!)
    const randomPassword = faker.internet.password()
    const success = await userService.setPassword(jsonBody.email, randomPassword)
    if (!success) {
      return ctx.json(null, 400)
    }
    return ctx.json({ temporaryPassword: randomPassword })
  }
)

accountsRoute.post(
  '/set-password',
  zValidator('json', setPasswordRequestSchema, (result, c) => {
    if (!result.success) {
      const errorMessage = result.error.issues.map((issue) => issue.message).join(', ')
      return c.render(<span className='text-red-700'>{errorMessage}</span>)
    }
  }),
  async (ctx) => {
    const request = await ctx.req.json()
    const userService = UserService(ctx.env.DB, ctx.env.AUTH_SECRET_KEY!)
    const user = await userService.getUserByEmail(request.email)
    if (!user) {
      return ctx.render(<span className='text-red-700'>User not found</span>)
    }

    const existingPassword = request.currentPassword
    const hashedPassword = hashPassword(existingPassword, ctx.env.AUTH_SECRET_KEY!)

    if (user.password !== hashedPassword) {
      return ctx.render(<span className='text-red-700'>Incorrect password</span>)
    }

    await userService.setPassword(user.email, request.newPassword)
    ctx.res.headers.set('HX-Retarget', 'this')
    return ctx.render(
      <span className='text-green-700 text-2xl'>Password updated successfully!</span>
    )
  }
)

export { accountsRoute }
