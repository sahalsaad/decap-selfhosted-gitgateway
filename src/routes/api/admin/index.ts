import { OpenAPIHono } from '@hono/zod-openapi'
import { defaultHook } from 'stoker/openapi'

import type { BaseAppBindings } from '@/types/app-bindings'

import accountRoutes from '@/src/routes/api/admin/account/account.routes'
import authRoutes from '@/src/routes/api/admin/auth/auth.routes'
import sitesRoute from '@/src/routes/api/admin/site/site.routes'
import userRoutes from '@/src/routes/api/admin/user/user.routes'

const adminRoutes = new OpenAPIHono<BaseAppBindings>({ defaultHook })

// admin routes
adminRoutes.route('/auth', authRoutes)
adminRoutes.route('/site', sitesRoute)
adminRoutes.route('/account', accountRoutes)
adminRoutes.route('/user', userRoutes)

export default adminRoutes
