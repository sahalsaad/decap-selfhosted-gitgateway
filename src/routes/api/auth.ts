import { createRoute, OpenAPIHono, z } from '@hono/zod-openapi'
import { basicAuth } from 'hono/basic-auth'
import { sign } from 'hono/jwt'
import { timingSafeEqual } from 'hono/utils/buffer'

import type { AppBindings } from '@/types/app-bindings'

import { hashPassword } from '@services/encryption-service'
import { UserService } from '@services/user-service'

const adminAuthRoute = new OpenAPIHono<AppBindings<BasicAuthVariables>>()

adminAuthRoute.use(
  '/',
  basicAuth({
    verifyUser: async (email: string, password: string, ctx) => {
      const userService = UserService(ctx.env.DB, ctx.env.AUTH_SECRET_KEY!)
      const user = await userService.getUserByEmail(email)
      if (!user) {
        return false
      }

      const hashedPassword = hashPassword(password, ctx.env.AUTH_SECRET_KEY)
      if (
        (await timingSafeEqual(user.password, hashedPassword))
        && user.role === 'admin'
      ) {
        ctx.set('user', {
          id: user.id,
          email: user.email,
          role: user.role,
        })
        return true
      }

      return false
    },
    invalidUserMessage: 'Invalid username or password or not admin',
  }),
)

const JwtResponseSchema = z.object({
  token_type: z.literal('bearer'),
  access_token: z.string(),
  expires_in: z.number(),
})

type JwtResponse = z.infer<typeof JwtResponseSchema>

const authRoute = createRoute({
  tags: ['Auth'],
  method: 'get',
  path: '/',
  responses: {
    401: {
      description: 'Invalid credentials',
    },
    200: {
      content: {
        'application/json': {
          schema: JwtResponseSchema,
        },
      },
      description: 'Retrieve the auth token',
    },
  },
  security: [{ basicAuth: [] }],
})

adminAuthRoute.openapi(authRoute, async (ctx) => {
  const duration = 60 * 60 * 12
  const user = ctx.get('user')
  const jwtPayload = {
    user: {
      id: user.id,
      email: user.email,
      role: user.role,
    },
    exp: Math.floor(Date.now() / 1000) + duration,
  }
  const accessToken = await sign(jwtPayload, ctx.env.AUTH_SECRET_KEY!)
  const jwt: JwtResponse = {
    token_type: 'bearer',
    access_token: accessToken,
    expires_in: duration * 1000,
  }

  return ctx.json(jwt, 200)
})

export { adminAuthRoute }
