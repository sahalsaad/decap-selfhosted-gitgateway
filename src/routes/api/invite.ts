import { zValidator } from '@hono/zod-validator'
import { Hono } from 'hono'

import type { JwtVariables } from '@/types/jwt-variables'

import { jwtAdminMiddleware, jwtMiddleware } from '@/src/middlewares/jwt'
import { inviteCreateRequestSchema } from '@/types/invite'
import { InviteService } from '@services/invite-service'

const inviteRoute = new Hono<{
  Bindings: CloudflareBindings
  Variables: JwtVariables
}>()

inviteRoute.use('/*', jwtMiddleware, jwtAdminMiddleware)
inviteRoute.post('/', zValidator('json', inviteCreateRequestSchema), async (ctx) => {
  const inviteRequest = ctx.req.valid('json')

  const inviteService = InviteService(ctx.env.DB)
  const result = await inviteService.createInvite(inviteRequest)

  const url = new URL(ctx.req.url)

  return ctx.json(
    {
      inviteUrl: `${url.origin}/register?invite=${result.id}`,
    },
    201,
  )
})

export { inviteRoute }
