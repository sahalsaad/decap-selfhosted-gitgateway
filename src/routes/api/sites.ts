import { zValidator } from '@hono/zod-validator'
import { Hono } from 'hono'

import type { AppBindings } from '@/types/app-bindings'
import type { JwtVariables } from '@/types/jwt-variables'

import { siteCreateRequestSchema, siteUpdateRequestSchema } from '@selfTypes/sites'
import { jwtAdminMiddleware, jwtMiddleware } from '@server/middlewares/jwt'
import { SiteService } from '@services/site-service'
import { UserService } from '@services/user-service'

const sitesRoute = new Hono<AppBindings<JwtVariables>>()

sitesRoute.use('/*', jwtMiddleware, jwtAdminMiddleware)
sitesRoute.post('/', zValidator('json', siteCreateRequestSchema), async (ctx) => {
  const { user } = ctx.get('jwtPayload')

  const siteCreateRequest = ctx.req.valid('json')
  const siteService = SiteService(ctx.env.DB, ctx.env.AUTH_SECRET_KEY!)
  const createdSite = await siteService.createSite(siteCreateRequest)
  const userService = UserService(ctx.env.DB, ctx.env.AUTH_SECRET_KEY!)
  await userService.addUserSite(user.id, createdSite.id)

  return ctx.json(createdSite, 201)
})

sitesRoute.put('/:siteId', zValidator('json', siteUpdateRequestSchema), async (ctx) => {
  const siteUpdateRequest = ctx.req.valid('json')
  const siteId = ctx.req.param('siteId')

  const siteService = SiteService(ctx.env.DB, ctx.env.AUTH_SECRET_KEY!)
  const isSuccess = await siteService.updateSite(siteId, siteUpdateRequest)

  if (!isSuccess) {
    return ctx.notFound()
  }

  return ctx.body(null, 204)
})

sitesRoute.delete('/:siteId', async (ctx) => {
  const siteId = ctx.req.param('siteId')
  const siteService = SiteService(ctx.env.DB, ctx.env.AUTH_SECRET_KEY!)
  const isSuccess = await siteService.deleteSite(siteId)

  if (isSuccess) {
    return ctx.body(null, 204)
  }

  return ctx.notFound()
})

sitesRoute.get('/:siteId', async (ctx) => {
  const siteId = ctx.req.param('siteId')
  const siteService = SiteService(ctx.env.DB, ctx.env.AUTH_SECRET_KEY!)
  const result = await siteService.getSiteById(siteId)

  if (!result) {
    return ctx.notFound()
  }

  return ctx.json(result)
})

sitesRoute.get('/', async (ctx) => {
  const siteService = SiteService(ctx.env.DB, ctx.env.AUTH_SECRET_KEY!)
  const result = await siteService.getAllSite()
  return ctx.json(result)
})

export { sitesRoute }
