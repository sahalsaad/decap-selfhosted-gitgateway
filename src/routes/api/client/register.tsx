import { zValidator } from '@hono/zod-validator'
import { Hono } from 'hono'

import type { BaseAppBindings } from '@/types/app-bindings'

import { ErrorMessage } from '@client/components/error-message'
import { SuccessMessage } from '@client/components/success-message'
import { inviteHandleRequestSchema } from '@selfTypes/invite'
import { InviteService } from '@services/invite-service'
import { SiteService } from '@services/site-service'
import { UserService } from '@services/user-service'

export default new Hono<BaseAppBindings>()
  .post('/', zValidator('json', inviteHandleRequestSchema, (result, ctx) => {
    if (!result.success) {
      const errorMessage = result.error.issues
        .map(issue => `${issue.path.join('.')}: ${issue.message}`)
        .join(', ')
      return ctx.render(<ErrorMessage>{errorMessage}</ErrorMessage>)
    }
  }), async (ctx) => {
    const createUserRequest = ctx.req.valid('json')

    const inviteService = InviteService(ctx.env.DB)
    const invite = await inviteService.getInviteById(createUserRequest.inviteId)

    if (!invite) {
      return ctx.render(<ErrorMessage>Invalid invite id</ErrorMessage>)
    }

    const deleteSuccess = await inviteService.deleteInvite(invite.id)
    if (!deleteSuccess) {
      return ctx.render(<ErrorMessage>Invalid invite id</ErrorMessage>)
    }

    const email = invite.email ? invite.email : createUserRequest.email

    const userService = UserService(ctx.env.DB, ctx.env.AUTH_SECRET_KEY!)
    const createdUser = await userService.createUser({
      email,
      firstName: createUserRequest.firstName,
      lastName: createUserRequest.lastName,
      password: createUserRequest.password,
      role: invite.role,
    })

    const successMessage = [
      <SuccessMessage>Register successful.</SuccessMessage>,
    ]

    if (invite.siteId) {
      const siteService = SiteService(ctx.env.DB, ctx.env.AUTH_SECRET_KEY!)
      await siteService.addUser(createdUser.id, invite.siteId)
      const site = await siteService.getSiteById(invite.siteId)
      successMessage.push(
        <a className="text-blue-500" href={site!.cmsUrl}>
          Go to CMS
        </a>,
      )
    }

    ctx.res.headers.set('HX-Retarget', 'this')
    return ctx.render(<div className="text-center">{successMessage}</div>)
  })
