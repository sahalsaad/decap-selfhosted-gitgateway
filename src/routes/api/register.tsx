import { Hono } from "hono";
import { InviteService } from "@services/invite-service";
import { UserService } from "@services/user-service";
import { SiteService } from "@services/site-service";
import { zValidator } from "@hono/zod-validator";
import { handleInviteSchema } from "@selfTypes/invite";

const registerRoute = new Hono<{ Bindings: CloudflareBindings }>();

registerRoute.post("/", zValidator("json", handleInviteSchema), async (ctx) => {
  const createUserRequest = ctx.req.valid("json");

  const inviteService = InviteService(ctx.env.DB);
  const invite = await inviteService.getInviteById(createUserRequest.inviteId);

  if (!invite) {
    return ctx.json(null, 400);
  }

  const email = invite.allowSetEmail ? createUserRequest.email : invite.email;

  const userService = UserService(ctx.env.DB, ctx.env.AUTH_SECRET_KEY!);
  const [createdUser] = await userService.createUser({
    email,
    firstName: createUserRequest.firstName,
    lastName: createUserRequest.lastName,
    password: createUserRequest.password,
    role: invite.role,
  });

  await inviteService.deleteInvite(invite.id);

  let successMessage = [<div>Register successful.</div>];

  if (invite.siteId) {
    await userService.addUserSite(createdUser.id, invite.siteId);
    const siteService = SiteService(ctx.env.DB, ctx.env.AUTH_SECRET_KEY!);
    const site = await siteService.getSiteById(invite.siteId);
    successMessage.push(
      <a className="text-blue-500" href={site!.url}>
        Go to CMS
      </a>,
    );
  }

  return ctx.render(<div className="text-center">{successMessage}</div>);
});

export { registerRoute };
