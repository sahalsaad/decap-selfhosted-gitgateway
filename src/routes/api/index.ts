import type { Variables } from '@selfTypes/variables'
import { registerRoute } from '@server/routes/api/register'
import { Hono } from 'hono'
import { adminAuthRoute } from './auth'
import { sitesRoute } from './sites'
import { usersRoute } from './users'

const apiRoute = new Hono<{
  Bindings: CloudflareBindings
  Variables: Variables
}>()

apiRoute.route('/sites', sitesRoute)
apiRoute.route('/users', usersRoute)
apiRoute.route('/auth', adminAuthRoute)
apiRoute.route('/register', registerRoute)

export { apiRoute }
