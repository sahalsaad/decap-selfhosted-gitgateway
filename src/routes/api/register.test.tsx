import { faker } from '@faker-js/faker'

import { registerRoute } from '@/src/routes/api/register'
import { generateSiteRequest, MOCK_ENV } from '@/vitest/data-helpers'

const mockInviteService = {
  getInviteById: vi.fn(),
  deleteInvite: vi.fn(),
}

const mockSiteService = {
  getSiteById: vi.fn(),
}

const mockUserService = {
  createUser: vi.fn(),
  addUserSite: vi.fn(),
}

vi.mock('@services/invite-service', () => ({
  InviteService: vi.fn().mockImplementation(() => mockInviteService),
}))

vi.mock('@services/site-service', () => ({
  SiteService: vi.fn().mockImplementation(() => mockSiteService),
}))

vi.mock('@services/user-service', () => ({
  UserService: vi.fn().mockImplementation(() => mockUserService),
}))

describe('register route', () => {
  afterEach(() => {
    vi.clearAllMocks()
  })
  it('should return 400 if invalid invite id', async () => {
    mockInviteService.getInviteById.mockResolvedValue(undefined)
    const response = await registerRoute.request(
      '/',
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          inviteId: faker.string.uuid(),
          email: faker.internet.email(),
          password: faker.internet.password(),
          firstName: faker.person.firstName(),
          lastName: faker.person.lastName(),
        }),
      },
      MOCK_ENV,
    )
    expect(response.status).toBe(400)
  })

  it('should return 200 if valid request', async () => {
    mockInviteService.getInviteById.mockResolvedValue({
      id: faker.string.uuid(),
      email: faker.internet.email(),
      role: faker.helpers.arrayElement(['admin', 'contributor']),
    })

    mockInviteService.deleteInvite.mockResolvedValue(true)

    const response = await registerRoute.request(
      '/',
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          inviteId: faker.string.uuid(),
          email: faker.internet.email(),
          password: faker.internet.password(),
          firstName: faker.person.firstName(),
          lastName: faker.person.lastName(),
        }),
      },
      MOCK_ENV,
    )
    expect(response.status).toBe(200)
    expect(await response.text()).toMatchSnapshot()
  })

  it('should add site link if siteId provided', async () => {
    mockInviteService.getInviteById.mockResolvedValue({
      id: faker.string.uuid(),
      email: faker.internet.email(),
      role: faker.helpers.arrayElement(['admin', 'contributor']),
      siteId: faker.string.uuid(),
    })

    mockInviteService.deleteInvite.mockResolvedValue(true)

    const mockSite = {
      ...generateSiteRequest(),
      cmsUrl: 'https://example.com',
    }
    mockSiteService.getSiteById.mockResolvedValue(mockSite)

    mockUserService.addUserSite.mockResolvedValue(true)
    mockUserService.createUser.mockResolvedValue({
      id: faker.string.uuid(),
    })

    const response = await registerRoute.request(
      '/',
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          inviteId: faker.string.uuid(),
          email: faker.internet.email(),
          password: faker.internet.password(),
          firstName: faker.person.firstName(),
          lastName: faker.person.lastName(),
          siteId: faker.string.uuid(),
        }),
      },
      MOCK_ENV,
    )
    expect(response.status).toBe(200)
    expect(await response.text()).toMatchSnapshot()
  })
})
