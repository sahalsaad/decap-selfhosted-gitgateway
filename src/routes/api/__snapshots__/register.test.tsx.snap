// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`register route > should add site link if siteId provided 1`] = `"<div class="text-center"><div>Register successful.</div><a class="text-blue-500" href="https://example.com">Go to CMS</a></div>"`;

exports[`register route > should return 200 if valid request 1`] = `"<div class="text-center"><div>Register successful.</div></div>"`;
